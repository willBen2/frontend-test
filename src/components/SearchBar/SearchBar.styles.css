.input-container {
    border: 1px solid cornflowerblue;
    border-top-left-radius: 10px;
    border-top-right-radius: 10px;
    overflow: hidden;
    padding-inline: 8px;
    padding-block: 4px;
}

.input {
    all: unset;
    width: 100%;
    outline: none;
    appearance: none;
    -webkit-appearance: none;
}
