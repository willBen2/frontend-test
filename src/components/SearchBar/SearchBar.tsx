import React, { ReactElement } from 'react';
// I prefer to use css modules, or less modules to avoid class colisions, but I'm keeping it simple
import './SearchBar.styles.css';
import { Input } from '~/components';
import { InputProps } from '~/components/Input/Input';

interface SearchBarProps {
  search?: string;
  children?: ReactElement;
  onSearchChange?: (search: string) => void;
}

interface ISearchComponent extends React.FC<SearchBarProps> {
  Input: typeof SearchInput;
}

interface SearchInput {
  children?: ReactElement;
  search?: string;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

function SearchInput(props: SearchInput & InputProps) {
  const { children, search, onChange, ...rest } = props;
  if (children && children.type === Input) {
    return React.cloneElement(children, {
      value: search,
      onChange,
      ...rest,
    });
  }
  return <Input className={props.className} value={search} onChange={onChange} {...rest} />;
}

const SearchBar: ISearchComponent = (props: SearchBarProps) => {
  const { onSearchChange = () => {}, search, children } = props;
  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    onSearchChange(event.target.value);
  };

  return (
    <div className="input-container">
      <SearchInput className="input" children={children} search={search} onChange={onChange} />
    </div>
  );
}

SearchBar.Input = Input;

export default SearchBar;
