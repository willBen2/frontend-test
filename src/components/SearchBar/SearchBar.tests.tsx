import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import SearchBar from './SearchBar';

describe('SearchBar', () => {
  it('should match with default SearchBar', () => {
    const { container } = render(
      <SearchBar />
    )
    expect(container).toMatchSnapshot();
  });

  it('should fire onSearchChange', () => {
    const onChangeMock = jest.fn();
    render(
      <SearchBar onSearchChange={onChangeMock}>
        <SearchBar.Input data-testid="input" />
      </SearchBar>
    );

    fireEvent.change(screen.getByTestId('input'), {
      target: { value: 'test' },
    });

    expect(onChangeMock).toHaveBeenCalled();
  });
})
