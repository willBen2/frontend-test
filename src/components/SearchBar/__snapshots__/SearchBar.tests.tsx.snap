// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should match with default SearchBar 1`] = `
<div>
  <div
    class="input-container"
  >
    <input
      class="input"
      value=""
    />
  </div>
</div>
`;
